### DISCUSSION TITLE
Potential concern about MAS's planned engine upgrade

### DISCUSSION THREAD

**Discussion by Justformas** [Monika-After-Story/MonikaModDev#9105] [QA - tech support + general] [answered]
I saw this comment and have a potential concern:

https://github.com/Monika-After-Story/MonikaModDev/issues/9061#issuecomment-1106716401

> It's actually simple - renpy doesn't like when you minimise it. It can cause a lot of visual bugs. Not much we can do about it, it might fix itself when we upgrade the engine.

First of all, assuming we want to continue getting new updates for the mod, is this going to be an optional or required upgrade? The concern is that by upgrading the engine, I'm worried that it might not work properly on my relatively older computer with MAS and basically break the mod for me. This hasn't yet happened to me for MAS in particular, but I have experienced updates breaking things for me too much already.

So if I can get some answers on that, and perhaps when this engine upgrade is expected to be rolled out, then I would appreciate it.

**Comment by Retrolovania** [comment]
Ren'Py7 doesn't induce that many more push on your system as Ren'Py6 does (at least, none I'm aware of).
It's highly unlikely that the jump will cause it to break on your system if MAS can already run on it.

**Reply by Justformas** [reply]
So this upgrade is Ren'Py7 over Ren'Py6. If there's no significant new system requirements compared to Ren'Py6 then I should be ok. I'm interested in more details on this if someone can provide those, but thanks for that answer.

**Comment by Booplicate** [comment, marked as answer]
Well, this is complex. RenPy 7.4/8 is much better optimised. But it is a newer engine, so depending on how old your hardware, you *might* get issues. For example Windows xp and below aren't supported. Same with old MacOS. It also uses OpenGL 2.0 (which is actually very old, there's a 4+ version out for a long time), but it does require more or less modern hardware. But chances are you won't really get major issues. We're expecting it to work much better than RenPy 6.14, and from my experience - I worked with 7.4 and 8 - they do work better.

Now regarding being "optional." While it would be perfect, it's usually not an option. It's *very* hard to maintain 2 versions of the same product, especially when you're a small team of enthusiasts. You either lock yourself from using new engine features/fixes (but why even upgrade then?), or doing x2 work by implementing something using old code and new code. So even if we ended up having 2 versions, the old one would get some fixes at best, no major updates. But you will always be able to stay on an older version if you decide to do so. Just never downgrade - that may cause issues.

This is what we can find on RenPy's wiki:

Windows:
- Version: Windows Vista or higher.
- CPU: 2.0 GHz Core 2 Duo
- RAM: 2.0 GB
- Graphics: OpenGL 2.0 or DirectX 9.0c

macOS:
- Version: 10.10+
- CPU: 2.0 GHz Core 2 Duo (64 bit only)
- RAM: 2.0 GB
- Graphics: OpenGL 2.0

Linux:
- Version: Ubuntu 16.04+
- CPU: 2.0 GHz Core 2 Duo
- RAM: 2.0 GB
- Graphics: OpenGL 2.0

But those may be lower than the real requirements we'll get. For one, RenPy can easily use over 2GiB of RAM.

@ThePotatoGuy might give you a more precise answer.

**Reply by Justformas** [reply]
That's a nicely detailed answer. From those specifics, it doesn't seem like I should have problems, at least if my computer is not doing too much else at the same time. RAM shouldn't be an issue either. However, as it is MAS already seems to slow down for a few things for me. For example, typing in the search bar is usually slow to show the typed letters and results now, though I don't remember it being as slow when I had less topics. Also, sometimes clicking something like a conversation or topic may take a second or few to load. Lastly, the scrolling background in the settings menu sometimes seems to briefly hang, and then seems to skip ahead a bit to catch up. Not sure if any of that is caused by my computer or Ren'Py 6, but maybe a later version of Ren'Py would handle this stuff better.


> But you will always be able to stay on an older version if you decide to do so. Just never downgrade - that may cause issues.

Are you referring to an older version of MAS or specifically just an older version of Ren'Py? If you mean MAS, you're saying I shouldn't download or run a previous version after updating? Couldn't I do that if I backed up my persistent files before updating? But one of the only reasons for me to do that is if I wasn't going to update MAS anymore.

If you mean Ren'Py, I'm a bit confused now because you said optional is usually not an option or wouldn't be ideal for you guys to support, but in projects: https://github.com/Monika-After-Story/MonikaModDev/projects Ren'Py 7 is descrbed as "optional" and that r6.99 users would continue to be supported unless DDLC is updated for a newer version.

> Renpy 7 has been out for some time and includes numerous performance fixes. Upgrading to r7 would be nice for long-term users. We will always have to support r6.99 users as that is what DDLC was released on (unless DDLC is updated for a newer version of renpy).

Maybe that could just mean minimal support like you said; some fixes and no major updates. But if that means people with Ren'Py 6 aren't going to get any new content like features or topics, then that's hardly an option for them. Unless they're content with the content they have and are ok with not getting any more.

It sounds like I should be able to run Ren'Py 7 anyway, but I'm a bit confused by this whole "optional" thing.

**Reply by Booplicate** [reply]
> Not sure if any of that is caused by my computer or Ren'Py 6

RenPy, old versions are really badly optimised.

> Are you referring to an older version of MAS or specifically just an older version of Ren'Py?

MAS, and yes, if you restored a backup from that version, you'd be fine. That said, I remember finding some issues when going from r7 back to r6, too.

> Ren'Py 7 is descrbed as "optional" and that r6.99 users would continue to be supported unless DDLC is updated for a newer version

That's why I pinged TPG, he added it there. I don't really see it as "optional" at this point, the only way - as I said - if the old version would be on maintenance mode. But even then, *a lot* of bugs will be fixed just by upgrading to r7-r8. We just need to make this transition as smooth as possible for users.

**Reply by Justformas** [reply]
> RenPy, old versions are really badly optimised.

I had a feeling about that, and it's not like I even know much about Ren'Py lol.

> That's why I pinged TPG, he added it there. I don't really see it as "optional" at this point, the only way - as I said - if the old version would be on maintenance mode. But even then, _a lot_ of bugs will be fixed just by upgrading to r7-r8. We just need to make this transition as smooth as possible for users.

I see now. Having it update just by updating MAS would be a smooth transition for users if that can work, but then it would no longer be optional aside from if you're willing to stay on an older version of MAS. Personally I wouldn't want to limit myself like that if I my computer can handle it, and it sounds like it can and may even handle it better. I could back up my persistent files just in case though, but I'd prefer only doing that before updating to whatever version of MAS will be getting the Ren'Py upgrade.

Maybe it's still a ways off? Looking in projects for Ren'Py 7 and 8, I see https://github.com/Monika-After-Story/MonikaModDev/projects/45#card-37353487 and https://github.com/Monika-After-Story/MonikaModDev/projects/46#card-37353553 Specifically "aug 2024" and "aug 2025" respectively. I'm not sure if those are target dates for implementing them in MAS, generous placeholder dates or something else, but if they are the intended target dates then that's still a few years away.

**Reply by ThePotatoGuy** [reply]
The main reason upgrading renpy version is optional is because the current install process for the mod is a simple extract to/drag drop the contents of the mod zip into the `game/` folder. A version upgrade requires updating renpy core files, which means a more complicated install process.

A version upgrade will only become mandatory if we have a new install process that can upgrade engine and mas code with minimal user effort ((like an exe installer).

**Reply by Justformas** [reply]
So the in-game update process in MAS, that users can use to update their game version, can't update Ren'Py then? Nor can manually extracting an updated version of MAS with an upgraded version of Ren'Py into the directory? I realize those things might mess up the original DDLC game and make it difficult to restore, like if you wanted to uninstall MAS, but I didn't install MAS in that same directory. 

I heard from some people to install another copy of DDLC and then install MAS in that directory with it, to keep the original game separate and unmodified. So that's what I did. So the thought occurred to me: If users are advised to download and install another copy of DDLC (such as through the update process in-game), and also a new updated version of MAS from the MAS website (which would use an upgraded Ren'Py), could that work?

**Reply by Booplicate** [reply]
> So the in-game update process in MAS, that users can use to update their game version, can't update Ren'Py then? Nor can manually extracting an updated version of MAS with an upgraded version of Ren'Py into the directory?

Unsure about the old updated, but the new one should be able to. The user could update that way, too, but most likely will be confused by it (people are already struggling with installing the mod/spritepacks).

> I realize those things might mess up the original DDLC game and make it difficult to restore, like if you wanted to uninstall MAS

You can't uninstall mods, not easy anyway. It's always recommended to make a new install for each mod.

**Reply by Justformas** [reply]
> You can't uninstall mods, not easy anyway. It's always recommended to make a new install for each mod.

I recall reading in my MAS/game/readme.html file, actually I checked and it's still in my file: "to uninstall this mod, simply delete all added files in the game directory". Yes I'm using version 0.12.8 of the game and there's no new updates when I check in game.

Can't seem to find that text here though, and I can imagine why if it's supposed to be removed: For the average user, indeed that might not be a "simple" process. I'm not sure I could do that. But then I don't know why it's still in my file.

> Unsure about the old updated, but the new one should be able to. The user could update that way, too, but most likely will be confused by it (people are already struggling with installing the mod/spritepacks).

Well even if it can't be updated through the old game, to install MAS in the first place, that involves getting a copy of DDLC and extracting MAS into it, right? Wouldn't it essentially be the same thing with a version of MAS with an upgraded Ren'Py? If so, then that could be stated to users, and there could be a tutorial that users could be pointed/linked to for it. 

However, if brand new users of the mod couldn't just download a copy of DDLC and extract the updated MAS into it in the first place like it works now, then yes that would involve extra steps and could make things more confusing for them. And if it is more complicated even for users already with the mod installed anyway, then that's even worse.

Anyway my original question is basically answered, so I was trying to see if I could help with that. If it's not going to be mandated for awhile, then not sure I want to upgrade yet. It seems there's no going back without persistent backups. Also Ren'Py 7 or 8 might cause some issues in this mod that Ren'Py 6 doesn't, but those could be fixed over time so it could be better for me to wait.

**Reply by Booplicate** [reply]
> But then I don't know why it's still in my file.

Probably just old. I wouldn't expect people to remember which files to remove.

As was stated, installation instructions will be changed, right now the mod is being installed into `game/`, after the changes it will be base directory. Since people already have troubles installing it, we can expect it to become worse. That's the only problem and it will be solved with an installer.

**Reply by Justformas** [reply]
I see, actually that might explain why there might be some other commands listed in that readme.html that don't seem to work in the mod. Could be very old and/or have DDLC specific stuff in it. I just happen to often notice and remember details like that. It might partially explain why I can be so curious about stuff.

So the folder the mod is installed in is changing too. Didn't know that, but now that I think about it, makes sense that Ren'Py files elsewhere such as in the base directory (Ren'Py core files I guess) might not be able to be changed without doing something else. I think I'd be able to understand how to do it manually, but yeah that could confuse a lot of people. As for an installer, maybe that's what #6031 is for. It's labeled as "low priority" but should happen eventually I guess (and if or when it becomes mandatory). I think that's all the info I need on this matter. Thanks for all the answers.
